identity/src/commonMain/kotlin/com/android/identity/util/ByteArrayUtil.kt

```diff
@@ -0,0 +1,234 @@
+    require(value in validRange) { "Value $value is out of Int8 range" }
+    this[offset] = value.toByte()
+}
+
+fun ByteArray.putInt16(offset: Int, value: Int, validRange: IntRange = Short.MIN_VALUE..Short.MAX_VALUE) {
+    require(value in validRange) { "Value $value is out of Int16 range" }
+    this[offset] = ((value shr 8) and 0xFF).toByte()  // Higher byte.
+    this[offset + 1] = (value and 0xFF).toByte() // Lowe byte.
+}
+
+fun ByteArray.putInt16Le(offset: Int, value: Int, validRange: IntRange = Short.MIN_VALUE..Short.MAX_VALUE) {
+    require(value in validRange) { "Value $value is out of Int16 range" }
+    this[offset] = (value and 0xFF).toByte() // Lower byte.
+    this[offset + 1] = ((value shr 8) and 0xFF).toByte()  // Higher byte.
+}
+
+fun ByteArray.putInt32(offset: Int, value: Int, validRange: IntRange = Int.MIN_VALUE..Int.MAX_VALUE) {
+    require(value in validRange) { "Value $value is out of Int32 range" }
+    this[offset] = ((value shr 24) and 0xFF).toByte()
+    this[offset + 1] = ((value shr 16) and 0xFF).toByte()
+    this[offset + 2] = ((value shr 8) and 0xFF).toByte()
+    this[offset + 3] = (value and 0xFF).toByte()
+}
+
+fun ByteArray.putInt32Le(offset: Int, value: Int, validRange: IntRange = Int.MIN_VALUE..Int.MAX_VALUE) {
+    require(value in validRange) { "Value $value is out of Int32 range" }
+    this[offset] = (value and 0xFF).toByte()
+    this[offset + 1] = ((value shr 8) and 0xFF).toByte()
+    this[offset + 2] = ((value shr 16) and 0xFF).toByte()
+    this[offset + 3] = ((value shr 24) and 0xFF).toByte()
+}
+
+fun ByteArray.putInt64(offset: Int, value: Long, validRange: LongRange = Long.MIN_VALUE..Long.MAX_VALUE) {
+    require(value in validRange) { "Value $value is out of Int64 range" }
+    this[offset] = ((value shr 56) and 0xFF).toByte()
+    this[offset + 1] = ((value shr 48) and 0xFF).toByte()
+    this[offset + 2] = ((value shr 40) and 0xFF).toByte()
+    this[offset + 3] = ((value shr 32) and 0xFF).toByte()
+    this[offset + 4] = ((value shr 24) and 0xFF).toByte()
+    this[offset + 5] = ((value shr 16) and 0xFF).toByte()
+    this[offset + 6] = ((value shr 8) and 0xFF).toByte()
+    this[offset + 7] = (value and 0xFF).toByte()
+}
+
+fun ByteArray.putInt64Le(offset: Int, value: Long, validRange: LongRange = Long.MIN_VALUE..Long.MAX_VALUE) {
+    require(value in validRange) { "Value $value is out of Int64 range" }
+    this[offset] = (value and 0xFF).toByte()
+    this[offset + 1] = ((value shr 8) and 0xFF).toByte()
+    this[offset + 2] = ((value shr 16) and 0xFF).toByte()
+    this[offset + 3] = ((value shr 24) and 0xFF).toByte()
+    this[offset + 4] = ((value shr 32) and 0xFF).toByte()
+    this[offset + 5] = ((value shr 40) and 0xFF).toByte()
+    this[offset + 6] = ((value shr 48) and 0xFF).toByte()
+    this[offset + 7] = ((value shr 56) and 0xFF).toByte()
+}
+
+fun ByteArray.putUInt8(offset: Int, value: Int, validRange: IntRange = 0..UByte.MAX_VALUE.toInt()) {
```

Comment:
I think we should take an UInt/ULong for all the putU* functions and similarly return an an UInt/ULong for the getU* functions.

Reply:
+1 to returning an unsigned type for the get functions.

For the put functions, one possible reason we might want to support both signed and unsigned versions of the parameter: if the caller does an explicit cast (toUInt()), it will silently convert negative values to large positive ones. However, if we want to consider negative values to be invalid (eg. if there's untrusted input, or size calculations), then having a version of the function take a signed value would also allow us to add a check and throw an error on negative values.